File: java/4_ObjectOrientedProgramming/Inheritance.java
/**
 * Inheritance.java
 * Demonstrates inheritance: parent-child relationships
 */

public class Inheritance {
    
    static class Animal {
        protected String name;
        protected int age;
        
        public Animal(String name, int age) {
            this.name = name;
            this.age = age;
        }
        
        public void eat() {
            System.out.println(name + " is eating.");
        }
        
        public void sleep() {
            System.out.println(name + " is sleeping.");
        }
    }
    
    static class Dog extends Animal {
        private String breed;
        
        public Dog(String name, int age, String breed) {
            super(name, age);
            this.breed = breed;
        }
        
        @Override
        public void eat() {
            System.out.println(name + " is eating dog food.");
        }
        
        public void bark() {
            System.out.println(name + " is barking: Woof! Woof!");
        }
    }
    
    static class Cat extends Animal {
        public Cat(String name, int age) {
            super(name, age);
        }
        
        @Override
        public void eat() {
            System.out.println(name + " is eating cat food.");
        }
        
        public void meow() {
            System.out.println(name + " is meowing: Meow!");
        }
    }
    
    public static void main(String[] args) {
        Dog dog = new Dog("Buddy", 5, "Golden Retriever");
        Cat cat = new Cat("Whiskers", 3);
        
        dog.eat();
        dog.bark();
        dog.sleep();
        
        cat.eat();
        cat.meow();
    }
}

File: python/4_ObjectOrientedProgramming/inheritance.py
#!/usr/bin/env python3
"""
inheritance.py
Demonstrates inheritance in Python

Author: Code Fusion
Version: 1.0
"""

class Animal:
    """Base class for animals."""
    
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def eat(self):
        print(f"{self.name} is eating.")
    
    def sleep(self):
        print(f"{self.name} is sleeping.")

class Dog(Animal):
    """Dog class inherits from Animal."""
    
    def __init__(self, name, age, breed):
        super().__init__(name, age)
        self.breed = breed
    
    def eat(self):
        print(f"{self.name} is eating dog food.")
    
    def bark(self):
        print(f"{self.name} is barking: Woof! Woof!")
    
    def __str__(self):
        return f"Dog(name='{self.name}', breed='{self.breed}')"

class Cat(Animal):
    """Cat class inherits from Animal."""
    
    def eat(self):
        print(f"{self.name} is eating cat food.")
    
    def meow(self):
        print(f"{self.name} is meowing: Meow!")

if __name__ == "__main__":
    print("=== Inheritance in Python ===")
    print()
    
    dog = Dog("Buddy", 5, "Golden Retriever")
    dog.eat()
    dog.bark()
    dog.sleep()
    print()
    
    cat = Cat("Whiskers", 3)
    cat.eat()
    cat.meow()
    cat.sleep()
    print()
    
    # Polymorphism
    print("=== Polymorphism ===")
    animals = [dog, cat]
    for animal in animals:
        animal.eat()
